rest/server_context.go

```diff
@@ -839,14 +868,74 @@ func (sc *ServerContext) _getOrAddDatabaseFromConfig(ctx context.Context, config
 		return dbcontext, nil
 	}
 
-	atomic.StoreUint32(&dbcontext.State, db.DBOnline)
-	_ = dbcontext.EventMgr.RaiseDBStateChangeEvent(dbName, "online", stateChangeMsg, &sc.Config.API.AdminInterface)
+	// If asyncOnline wasn't specified, block until db init is completed, then start online processes
+	if !options.asyncOnline {
+		base.DebugfCtx(ctx, base.KeyConfig, "Waiting for database init to complete...")
+		if dbInitDoneChan != nil {
+			initError := <-dbInitDoneChan
+			if initError != nil {
+				return nil, initError
+			}
+		}
+		base.DebugfCtx(ctx, base.KeyConfig, "Database init completed, starting online processes")
+		if err := dbcontext.StartOnlineProcesses(ctx); err != nil {
+			return nil, err
+		}
+		atomic.StoreUint32(&dbcontext.State, db.DBOnline)
+		_ = dbcontext.EventMgr.RaiseDBStateChangeEvent(dbName, "online", stateChangeMsg, &sc.Config.API.AdminInterface)
+		return dbcontext, nil
+	} else {
+		// If asyncOnline is requested, set state to Starting and spawn a separate goroutine to wait for init completion
+		// before going online
+		base.DebugfCtx(ctx, base.KeyConfig, "Waiting for database init to complete asynchonously...")
+		atomic.StoreUint32(&dbcontext.State, db.DBStarting)
+		nonCancelCtx := base.NewNonCancelCtxForDatabase(dbName)
+		go sc.asyncDatabaseOnline(nonCancelCtx, dbcontext, dbInitDoneChan, config.Version)
+		return dbcontext, nil
+	}
+}
 
-	if err := dbcontext.StartOnlineProcesses(ctx); err != nil {
-		return nil, err
+// asyncDatabaseOnline waits for async initialization to complete (based on doneChan).  On successful completion, brings the database online.
+// Checks to ensure the database config hasn't been updated while waiting for init to complete - if that happens, doesn't attempt to bring
+// db online (it may have been set to offline=true)
+func (sc *ServerContext) asyncDatabaseOnline(nonCancelCtx base.NonCancellableContext, dbc *db.DatabaseContext, doneChan chan error, version string) {
+
+	ctx := nonCancelCtx.Ctx
+	if doneChan != nil {
+		initError := <-doneChan
+		if initError != nil {
+			base.WarnfCtx(ctx, "Async database init returned error: %v", initError)
+			return
+		}
+	}
+
+	// Before bringing the database online, ensure that the database hasn't been modified while we waited for initialization to complete
+	currentDbVersion := sc.GetDbVersion(dbc.Name)
+	if currentDbVersion != version {
+		base.InfofCtx(ctx, base.KeyConfig, "Database version changed while waiting for aync init - cancelling obsolete online request. Old version: %s New version: %s", version, currentDbVersion)
```

Comment:
```suggestion
		base.InfofCtx(ctx, base.KeyConfig, "Database version changed while waiting for async init - cancelling obsolete online request. Old version: %s New version: %s", version, currentDbVersion)
```

We possibly need to mark a state change on the db here too.